types.go

```diff
@@ -15,33 +15,39 @@ import (
 
 const defaultGradientAbsTol = 1e-6
 
-// EvaluationType is used by the optimizer to specify information needed
-// from the objective function.
-type EvaluationType int
+// EvaluationType is used by the optimizer to specify information needed from
+// the objective function. The types can be composed together using the binary
+// or operator.
+type EvaluationType uint
 
+// Basic evaluation types.
 const (
-	NoEvaluation EvaluationType = iota
-	FuncEvaluation
+	NoEvaluation   EvaluationType = 0
+	FuncEvaluation EvaluationType = 1 << iota
 	GradEvaluation
 	HessEvaluation
-	FuncGradEvaluation
-	FuncGradHessEvaluation
+)
+
+// Combined evaluation types.
```

Comment:
This should stay in this CL to keep the changes small, but I think we should remove these. It's easy to write in the optimizers, will make it clearer to new users how to compose the evaluations (by seeing the example in our code) and keeps the API surface smaller.